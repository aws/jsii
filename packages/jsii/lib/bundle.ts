import * as fs from 'fs-extra';
import { SPEC_FILE_NAME } from 'jsii-spec';
import * as os from 'os';
import * as path from 'path';
import * as webpack from 'webpack';
import { compilePackage } from './compiler';
import { normalizeJsiiModuleName } from './naming';
import readPackageMetadata from './package-metadata';
import { filterEmpty } from './util';

export async function bundle(moduleRoot: string) {
    const metadata = await readPackageMetadata(moduleRoot);
    const outputDir = metadata.outdir;
    const entrypoint = metadata.main;

    // make sure output directory is created
    await fs.mkdirs(outputDir);

    // compile the jsii spec and write to .jsii file
    const spec = await compilePackage(moduleRoot);

    const moduleName = spec.name;

    if (spec.typecount === 0) {
        // no exported types - do not create a jsii spec
        return outputDir;
    }

    const outputFile = await createSourceBundle();
    try {
        // add code to assembly.
        spec.code = await fs.readFile(outputFile, { encoding: 'utf-8' });
    } finally {
        await fs.unlink(outputFile);
    }

    const specOutputFilePath = path.join(outputDir, SPEC_FILE_NAME);
    await fs.writeJson(specOutputFilePath, spec, { replacer: filterEmpty, spaces: 2 });

    return outputDir;

    async function createSourceBundle(): Promise<string> {
        const bundleFileName = 'bundle.js';
        const externals: { [name: string]: string } = { };
        const webpackConfig: webpack.Configuration = {
            devtool: 'inline-source-map',
            entry: entrypoint,
            externals,
            context: path.resolve(path.join(__dirname, '..')),
            module: {
                rules: [
                    {
                        test: /\.js$/,
                        use: ['source-map-loader'],
                        enforce: 'pre'
                    }
                ]
            },
            node: {
                __dirname: true,
                __filename: true,
                Buffer: false,
                console: false,
                global: false,
                process: false,
                setImmediate: false,
            },
            output: {
                path: await fs.mkdtemp(os.tmpdir() + path.sep),
                filename: bundleFileName,
                library: moduleName,
                libraryTarget: 'var',
            },
            target: 'node'
        };

        // add all jsii packages as 'externals', which means they will not be bundled.
        for (const dep in spec.dependencies) {
            if (dep in spec.bundled) {
                continue; // bundle me!
            }

            const originalName = spec.dependencies[dep].package;
            externals[originalName] = normalizeJsiiModuleName(originalName);
        }

        return await new Promise<string>((resolve, reject) => {
            webpack(webpackConfig, (err, stats) => {
                if (err) {
                    reject(err);
                } else {
                    if (stats.hasErrors()) {
                        // tslint:disable-next-line:no-console
                        console.error(stats.toString());
                        process.exit(1);
                    }
                    resolve(path.join(webpackConfig.output!.path!, bundleFileName));
                }
            });
        });
    }
}
