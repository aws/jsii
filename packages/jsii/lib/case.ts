import * as Case from 'case';

const cached =
  (func: (text: string) => string): ((text: string) => string) =>
  (text: string) =>
    Cache.fetch(text, func);

export const camel = cached(Case.camel);
export const constant = cached(Case.constant);
export const pascal = cached(Case.pascal);
export const snake = cached(Case.snake);

class Cache {
  // Cache is indexed on a weak CacheKey so the cache can be purged under memory pressure
  private static readonly CACHES = new WeakMap<CacheKey, Map<string, string>>();

  public static fetch(text: string, func: (text: string) => string): string {
    // Check whether we have a cache for this function...
    const cacheKey = CacheKey.for(func);
    let cache = this.CACHES.get(cacheKey);
    if (cache == null) {
      // If not, create one...
      cache = new Map<string, string>();
      this.CACHES.set(cacheKey, cache);
    }

    // Check if the current cache has a value for this text...
    const cached = cache.get(text);
    if (cached != null) {
      return cached;
    }

    // If not, compute one...
    const result = func(text);
    cache.set(text, result);
    return result;
  }

  private constructor() {}
}

class CacheKey {
  // Storing cache keys as weak references to allow garbage collection if there is memory pressure.
  private static readonly STORE = new Map<any, WeakRef<CacheKey>>();

  public static for(data: any) {
    const entry = this.STORE.get(data)?.deref();
    if (entry != null) {
      return entry;
    }
    const newKey = new CacheKey();
    this.STORE.set(data, new WeakRef(newKey));
    return newKey;
  }

  private constructor() {}
}
