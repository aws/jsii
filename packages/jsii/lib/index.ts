export * from './jsii-diagnostic';
export * from './symbol-id';
export * from './helpers';
