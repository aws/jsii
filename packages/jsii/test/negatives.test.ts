import fs = require('fs-extra');
import path = require('path');
import ts = require('typescript');
import { Compiler } from '../lib/compiler';
import { ProjectInfo } from '../lib/project-info';

const SOURCE_DIR = path.join(__dirname, 'negatives');

for (const source of fs.readdirSync(SOURCE_DIR)) {
    if (!source.startsWith('neg.') || !source.endsWith('.ts') || source.endsWith('.d.ts')) { continue; }
    const filePath = path.join(SOURCE_DIR, source);
    test(source.replace(/neg\.(.+)\.ts/, '$1'), async () => {
        const [expectations, strict] = await _getExpectedErrorMessage(filePath);
        expect(expectations.length, `Expected error messages should be specified using ${MATCH_ERROR_MARKER}`)
            .toBeGreaterThan(0);
        const compiler = new Compiler({ projectInfo: _makeProjectInfo(source), watch: false, failOnWarnings: strict });
        const emitResult = await compiler.emit(path.join(SOURCE_DIR, source));
        expect(emitResult.emitSkipped).toBeTruthy();
        const errors = emitResult.diagnostics.filter(diag =>
            diag.category === ts.DiagnosticCategory.Error
            || (strict && diag.category === ts.DiagnosticCategory.Warning));
        for (const expectation of expectations) {
            expect(
                errors.find(e => _messageText(e).indexOf(expectation) !== -1),
                `No error contained: ${expectation}. Errors: \n${errors.map((e, i) => `[${i}] ${e.messageText}`).join('\n')}`
            ).toBeDefined();
        }

        // Cleaning up...
        for (const file of await fs.readdir(SOURCE_DIR)) {
            if (file.startsWith('neg.') && (file.endsWith('.d.ts') || file.endsWith('.js'))) {
                await fs.remove(path.join(SOURCE_DIR, file));
            }
            await fs.remove(path.join(SOURCE_DIR, '.jsii'));
            await fs.remove(path.join(SOURCE_DIR, 'tsconfig.json'));
        }
    }, 10000);
}

const MATCH_ERROR_MARKER = '///!MATCH_ERROR:';
const STRICT_MARKER = '///!STRICT!';
async function _getExpectedErrorMessage(file: string): Promise<[string[], boolean]> {
    const data = await fs.readFile(file, { encoding: 'utf8' });
    const lines = data.split('\n');
    const matches = lines.filter(line => line.startsWith(MATCH_ERROR_MARKER))
                         .map(line => line.substr(MATCH_ERROR_MARKER.length).trim());
    const strict = lines.some(line => line.startsWith(STRICT_MARKER));
    return [matches, strict];
}

function _messageText(diagnostic: ts.Diagnostic | ts.DiagnosticMessageChain): string {
    if (typeof diagnostic.messageText === 'string') {
        return diagnostic.messageText;
    }
    if (diagnostic.messageText.next) {
        return diagnostic.messageText.messageText + '|' + _messageText(diagnostic.messageText.next[0]);
    }
    return diagnostic.messageText.messageText;
}

function _makeProjectInfo(types: string): ProjectInfo {
    return {
        projectRoot: SOURCE_DIR,
        packageJson: undefined,
        types,
        main: types.replace(/(?:\.d)?\.ts(x?)/, '.js$1'),
        name: 'jsii', // That's what package.json would tell if we look up...
        version: '0.0.1',
        jsiiVersionFormat: 'short',
        license: 'Apache-2.0',
        author: { name: 'John Doe', roles: ['author'] },
        repository: { type: 'git', url: 'https://github.com/aws/jsii.git' },
        dependencies: [],
        peerDependencies: [],
        transitiveDependencies: [],
        bundleDependencies: {},
        targets: {},
        excludeTypescript: [],
    };
}
