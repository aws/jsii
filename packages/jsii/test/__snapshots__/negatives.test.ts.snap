// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`behavior-requires-iprefix 1`] = `
neg.behavior-requires-iprefix.ts:1:18 - error JSII8007: Interface contains behavior. Rename "Something" to "ISomething"

1 export interface Something {
                   ~~~~~~~~~


`;

exports[`class-name 1`] = `
error JSII8000: Type names must be CamelCased. Rename "myclass" to "Myclass"

`;

exports[`class-name.1 1`] = `
error JSII8000: Type names must be CamelCased. Rename "My_class" to "MyClass"

`;

exports[`compilation-error 1`] = `
neg.compilation-error.ts:1:1 - error TS2304: Cannot find name 'boom'.

1 boom! > CompilerErrorIsHere;
  ~~~~
neg.compilation-error.ts:1:9 - error TS2304: Cannot find name 'CompilerErrorIsHere'.

1 boom! > CompilerErrorIsHere;
          ~~~~~~~~~~~~~~~~~~~

`;

exports[`const-enum 1`] = `
neg.const-enum.ts:1:8 - error JSII1000: Exported "const enum" declarations are not allowed

1 export const enum NotAllowed {
         ~~~~~


`;

exports[`double-interface-members 1`] = `
neg.double-interface-members.ts:2:3 - error JSII3008: The "foo" property of struct "jsii.A" must be "readonly". Rename "jsii.A" to "IA" if it is meant to be a behavioral interface.

2   foo: number;
    ~~~

neg.double-interface-members.ts:4:1 - error JSII5015: Interface "jsii.B" re-declares member "foo". This is not supported as it results in invalid C#.

4 export interface B extends A {
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
5   foo: number;
  ~~~~~~~~~~~~~~
6 }
  ~

neg.double-interface-members.ts:5:3 - error JSII3008: The "foo" property of struct "jsii.B" must be "readonly". Rename "jsii.B" to "IB" if it is meant to be a behavioral interface.

5   foo: number;
    ~~~


`;

exports[`double-interface-members-deeper 1`] = `
neg.double-interface-members-deeper.ts:9:1 - error JSII5015: Interface "jsii.IC" re-declares member "foo". This is not supported as it results in invalid C#.

 9 export interface IC extends IB {
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10   foo(): void;
   ~~~~~~~~~~~~~~
11 }
   ~


`;

exports[`double-interface-members-method 1`] = `
neg.double-interface-members-method.ts:4:1 - error JSII5015: Interface "jsii.IB" re-declares member "foo". This is not supported as it results in invalid C#.

4 export interface IB extends IA {
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
5   foo(): void;
  ~~~~~~~~~~~~~~
6 }
  ~


`;

exports[`downgrade-to-readonly 1`] = `
error JSII5010: "jsii.Implementation#property" turns mutable when implementing jsii.IInterface. Make it readonly

`;

exports[`enum-members.1 1`] = `
error JSII8001: Enum members must be ALL_CAPS. Rename "jsii.MyEnum.Goo" to "GOO"

`;

exports[`enum-name.1 1`] = `
error JSII8000: Type names must be CamelCased. Rename "myEnum" to "MyEnum"

`;

exports[`enum-name.2 1`] = `
error JSII8000: Type names must be CamelCased. Rename "My_Enum" to "MyEnum"

`;

exports[`expose-unexported-type-external 1`] = `
error JSII3000: Exported APIs cannot use un-exported type "jsii.UnexportedType"

`;

exports[`expose-unexported-type-internal 1`] = `
neg.expose-unexported-type-internal.ts:7:10 - error JSII3001: Type "UnexportedType" cannot be used as the property type because it is private or @internal

7   public p?: UnexportedType;
           ~

  neg.expose-unexported-type-internal.ts:4:1
    4 class UnexportedType {}
      ~~~~~~~~~~~~~~~~~~~~~~~
    The referenced type is declared here

`;

exports[`expose-unexported-type-internal-in-namespace 1`] = `
neg.expose-unexported-type-internal-in-namespace.ts:9:10 - error JSII3001: Type "UnexportedType" cannot be used as the property type because it is private or @internal

9   public p?: MyNamespace.UnexportedType;
           ~

  neg.expose-unexported-type-internal-in-namespace.ts:5:3
    5   export class UnexportedType {}
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    The referenced type is declared here

`;

exports[`expose-unexported-type-this 1`] = `
neg.expose-unexported-type-this.ts:10:38 - warning JSII5018: "boolean" is a reserved word in Java. Using this name may cause problems when generating language bindings. Consider a different name.

10   public constructor(public readonly boolean = true) {
                                        ~~~~~~~

neg.expose-unexported-type-this.ts:4:10 - error JSII3001: Type "this" (aka: "HiddenBaseClass") cannot be used as the return type because it is private or @internal

4   public returnsThis() {
           ~~~~~~~~~~~

  neg.expose-unexported-type-this.ts:3:1
      3 abstract class HiddenBaseClass {
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      4   public returnsThis() {
        ~~~~~~~~~~~~~~~~~~~~~~~~
    ... 
      6   }
        ~~~
      7 }
        ~
    The referenced type is declared here

`;

exports[`extend-struct 1`] = `
neg.extend-struct.ts:6:1 - error JSII3007: Attempt to extend or implement struct "jsii.Struct" from "jsii.IIllegal"

6 export interface IIllegal extends Struct {
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
7   method(): void;
  ~~~~~~~~~~~~~~~~~
8 }
  ~


`;

exports[`implement-struct 1`] = `
neg.implement-struct.ts:6:1 - error JSII3007: Attempt to extend or implement struct "jsii.Struct" from "jsii.Illegal"

  6 export class Illegal implements Struct {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  7   public readonly field: string = 'foo';
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
... 
 11   }
    ~~~
 12 }
    ~


`;

exports[`implementation-changes-types.1 1`] = `
error JSII5003: "jsii.Something#returnSomething" changes the return type to "jsii.Subclass" when implementing jsii.ISomething. Change it to "jsii.Superclass"

`;

exports[`implementation-changes-types.2 1`] = `
error JSII5003: "jsii.ISomethingElse#returnSomething" changes the return type to "jsii.Subclass" when implementing jsii.ISomething. Change it to "jsii.Superclass"

`;

exports[`implementation-changes-types.3 1`] = `
error JSII5006: "jsii.Something#takeSomething" changes the type of parameter "_argument" to jsii.Subclass when implementing jsii.ISomething. Change it to jsii.Superclass

`;

exports[`implementation-changes-types.4 1`] = `
error JSII5004: "jsii.SomethingImpl#something" changes the property type to "jsii.Subclass" when implementing jsii.ISomething. Change it to "jsii.Superclass"

`;

exports[`implementation-changes-types.5 1`] = `
error JSII5004: "jsii.ISomethingElse#something" changes the property type to "jsii.Subclass" when implementing jsii.ISomething. Change it to "jsii.Superclass"

`;

exports[`implementing-method-changes-optionality 1`] = `
error JSII5008: "jsii.Implementor#method" turns parameter "_optional" required when implementing jsii.IInterface. Make it optional

`;

exports[`implementing-method-changes-optionality.1 1`] = `
error JSII5008: "jsii.Implementor#method" turns parameter "_optional" required when overriding jsii.AbstractClass. Make it optional

`;

exports[`implementing-method-changes-optionality.2 1`] = `
error JSII5008: "jsii.Implementor#method" turns parameter "_optional" required when overriding jsii.ParentClass. Make it optional

`;

exports[`implementing-property-changes-optionality 1`] = `
error JSII5009: "jsii.Implementor#property" turns  required when implementing jsii.IInterface. Make it optional

`;

exports[`implementing-property-changes-optionality.1 1`] = `
error JSII5009: "jsii.Implementor#property" turns  required when overriding jsii.AbstractClass. Make it optional

`;

exports[`implementing-property-changes-optionality.2 1`] = `
error JSII5009: "jsii.Implementor#property" turns  required when overriding jsii.ParentClass. Make it optional

`;

exports[`implements-class 1`] = `
neg.implements-class.ts:1:1 - error JSII3005: Type "jsii.NotAnInterface" cannot be used as an interface

  1 export class NotAnInterface {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  2   public meaningOfTheUniverse() {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
... 
  4   }
    ~~~
  5 }
    ~


`;

exports[`inheritance-changes-types.1 1`] = `
error JSII5003: "jsii.SomethingSpecific#returnSomething" changes the return type to "jsii.Subclass" when overriding jsii.Something. Change it to "jsii.Superclass"

`;

exports[`inheritance-changes-types.2 1`] = `
error JSII5003: "jsii.SomethingSpecific#returnSomething" changes the return type to "jsii.Subclass" when overriding jsii.Something. Change it to "jsii.Superclass"

`;

exports[`inheritance-changes-types.3 1`] = `
error JSII5006: "jsii.SomethingSpecific#takeSomething" changes the type of parameter "_argument" to jsii.Subclass when overriding jsii.Something. Change it to jsii.Superclass

`;

exports[`inheritance-changes-types.4 1`] = `
error JSII5004: "jsii.SomethingSpecific#something" changes the property type to "jsii.Subclass" when overriding jsii.SomethingUnspecific. Change it to "jsii.Superclass"

`;

exports[`inheritance-changes-types.5 1`] = `
error JSII5004: "jsii.SomethingElse#something" changes the property type to "jsii.Subclass" when overriding jsii.SomethingBase. Change it to "jsii.Superclass"

`;

exports[`internal-underscore-class.5 1`] = `
neg.internal-underscore-class.5.ts:3:10 - error JSII8005: Members marked with @internal must have a name starting with "_". Rename "propertyWithInternalButNotUnderscorePrefix" to "_propertyWithInternalButNotUnderscorePrefix"

3   public propertyWithInternalButNotUnderscorePrefix?: string;
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


`;

exports[`internal-underscore-class.6 1`] = `
neg.internal-underscore-class.6.ts:2:10 - error JSII8006: Members with a name starting with "_" (e.g: "_propertyWithUnderscoreButNoInternal") must be marked @internal

2   public _propertyWithUnderscoreButNoInternal?: string;
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


`;

exports[`internal-underscore-class.7 1`] = `
neg.internal-underscore-class.7.ts:3:10 - error JSII8005: Members marked with @internal must have a name starting with "_". Rename "methodWithInternalButNoUnderscore" to "_methodWithInternalButNoUnderscore"

3   public methodWithInternalButNoUnderscore(): string {
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


`;

exports[`internal-underscore-class.8 1`] = `
neg.internal-underscore-class.8.ts:2:3 - error JSII8006: Members with a name starting with "_" (e.g: "_methodWithUnderscoreButNoInternal") must be marked @internal

2   _methodWithUnderscoreButNoInternal(): void;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


`;

exports[`internal-underscore-interface.1 1`] = `
neg.internal-underscore-interface.1.ts:3:3 - error JSII8005: Members marked with @internal must have a name starting with "_". Rename "propertyWithInternalButNotUnderscorePrefix" to "_propertyWithInternalButNotUnderscorePrefix"

3   propertyWithInternalButNotUnderscorePrefix: string;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


`;

exports[`internal-underscore-interface.2 1`] = `
neg.internal-underscore-interface.2.ts:2:3 - error JSII8006: Members with a name starting with "_" (e.g: "_propertyWithUnderscoreButNoInternal") must be marked @internal

2   _propertyWithUnderscoreButNoInternal: string;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


`;

exports[`internal-underscore-interface.3 1`] = `
neg.internal-underscore-interface.3.ts:3:3 - error JSII8005: Members marked with @internal must have a name starting with "_". Rename "methodWithInternalButNoUnderscore" to "_methodWithInternalButNoUnderscore"

3   methodWithInternalButNoUnderscore(): string;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


`;

exports[`internal-underscore-interface.4 1`] = `
neg.internal-underscore-interface.4.ts:2:10 - error JSII8006: Members with a name starting with "_" (e.g: "_methodWithUnderscoreButNoInternal") must be marked @internal

2   public _methodWithUnderscoreButNoInternal() {
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


`;

exports[`member-named-after-type 1`] = `
neg.member-named-after-type.ts:14:12 - warning JSII5019: The property name "myStruct" conflicts with the declaring interface "MyStruct". This will result in renaming the interface to "_MyStruct" in C#. Consider renaming "myStruct".

14   readonly myStruct: string;
              ~~~~~~~~

  neg.member-named-after-type.ts:13:18
    13 export interface MyStruct {
                        ~~~~~~~~
    The declaring interface is introduced here
neg.member-named-after-type.ts:4:19 - warning JSII5019: The property name "typeName" conflicts with the declaring class "TypeName". This will result in renaming the class to "_TypeName" in C#. Consider renaming "typeName".

4   public readonly typeName = 1337;
                    ~~~~~~~~

  neg.member-named-after-type.ts:3:14
    3 export class TypeName {
                   ~~~~~~~~
    The declaring class is introduced here
neg.member-named-after-type.ts:8:10 - warning JSII5019: The method name "otherType" conflicts with the declaring class "OtherType". This will result in renaming the class to "_OtherType" in C#. Consider renaming "otherType".

8   public otherType() {
           ~~~~~~~~~

  neg.member-named-after-type.ts:7:14
    7 export class OtherType {
                   ~~~~~~~~~
    The declaring class is introduced here

`;

exports[`method-name.1 1`] = `
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "jsii.MyClass.METHOD" to "method"

`;

exports[`method-name.2 1`] = `
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "jsii.MyClass.hello_world" to "helloWorld"

`;

exports[`method-name.3 1`] = `
error JSII5000: Methods and properties cannot have names like "getXxx": those conflict with Java property getters. Rename "MyClass.getFoo"

`;

exports[`method-name.4 1`] = `
error JSII5001: Methods and properties cannot have names like "setXxx": those conflict with Java property setters. Rename "MyClass.setFoo"

`;

exports[`mix-datatype-and-arg-name 1`] = `
neg.mix-datatype-and-arg-name.ts:10:3 - error JSII5017: Parameter name "dontWorry" is also the name of a property in a struct parameter. Rename the positional parameter.

10   public dance(dontWorry: string, lyrics: Lyrics) {
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
11     return \`\${dontWorry}: \${lyrics.beHappy}\`;
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12   }
   ~~~


`;

exports[`mutable-datatype 1`] = `
neg.mutable-datatype.ts:3:3 - error JSII3008: The "notOkay" property of struct "jsii.DataType" must be "readonly". Rename "jsii.DataType" to "IDataType" if it is meant to be a behavioral interface.

3   notOkay: number; // properties should be "readonly"
    ~~~~~~~


`;

exports[`non-optional-after-optional-ctor 1`] = `
neg.non-optional-after-optional-ctor.ts:2:3 - error JSII3009: Parameter "_arg2" cannot be optional, as it precedes required parameter "_arg3"

2   constructor(_arg1: string, _arg2 = 'hello', _arg3: string) {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3     return;
  ~~~~~~~~~~~
4   }
  ~~~


`;

exports[`non-optional-after-optional-method 1`] = `
neg.non-optional-after-optional-method.ts:2:3 - error JSII3009: Parameter "_arg2" cannot be optional, as it precedes required parameter "_argX"

2   public foo(_arg1: string, _arg2 = 'hello', _argX: string, _arg4?: boolean) {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3     return;
  ~~~~~~~~~~~
4   }
  ~~~


`;

exports[`omit.1 1`] = `
neg.omit.1.ts:7:33 - error JSII3004: Illegal extends clause for an exported API: MappedType

7 export interface BarBaz extends Omit<FooBar, 'foo'> {
                                  ~~~~~~~~~~~~~~~~~~~

  ../../../../node_modules/typescript/lib/lib.es5.d.ts:1462:35
    1462 type Pick<T, K extends keyof T> = {
                                           ~
    1463     [P in K]: T[P];
         ~~~~~~~~~~~~~~~~~~~
    1464 };
         ~
    The invalid super type is declared here.

`;

exports[`omit.2 1`] = `
neg.omit.2.ts:7:32 - error JSII3004: Illegal implements clause for an exported API: MappedType

7 export class BarBaz implements Omit<FooBar, 'foo'> {
                                 ~~~~~~~~~~~~~~~~~~~

  ../../../../node_modules/typescript/lib/lib.es5.d.ts:1462:35
    1462 type Pick<T, K extends keyof T> = {
                                           ~
    1463     [P in K]: T[P];
         ~~~~~~~~~~~~~~~~~~~
    1464 };
         ~
    The invalid super type is declared here.

`;

exports[`omit.3 1`] = `
neg.omit.3.ts:8:3 - error JSII1003: Only string-indexed map types are supported

8   bar(): Omit<FooBar, 'foo'>;
    ~~~


`;

exports[`omit.4 1`] = `
neg.omit.4.ts:8:7 - error JSII1003: Only string-indexed map types are supported

8   bar(opts: Omit<FooBar, 'foo'>): void;
        ~~~~


`;

exports[`override-changes-visibility 1`] = `
error JSII5002: "jsii.ChildClass#method" changes visibility to public when overriding jsii.BaseClass. Change it to protected
error JSII5002: "jsii.ChildClass#property" changes visibility to public when overriding jsii.BaseClass. Change it to protected

`;

exports[`prohibited-member-name 1`] = `
neg.prohibited-member-name.ts:4:10 - error JSII5016: Members cannot be named "equals" as it conflicts with synthetic declarations in some languages.

4   public equals(): boolean {
           ~~~~~~


`;

exports[`property-name.1 1`] = `
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "jsii.MyClass.PROP" to "prop"

`;

exports[`property-name.2 1`] = `
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "jsii.MyClass.my_Prop" to "myProp"

`;

exports[`property-name.3 1`] = `
error JSII5000: Methods and properties cannot have names like "getXxx": those conflict with Java property getters. Rename "MyClass.getFoo"

`;

exports[`reserved.emits-warning 1`] = `
neg.reserved.emits-warning.ts:2:14 - warning JSII5018: "None" is a reserved word in Python. Using this name may cause problems when generating language bindings. Consider a different name.

2 export class None {
               ~~~~

neg.reserved.emits-warning.ts:3:19 - warning JSII5018: "do" is a reserved word in C#, Java. Using this name may cause problems when generating language bindings. Consider a different name.

3   public readonly do: boolean = true;
                    ~~

neg.reserved.emits-warning.ts:5:10 - warning JSII5018: "assert" is a reserved word in Java, Python. Using this name may cause problems when generating language bindings. Consider a different name.

5   public assert(_internal: boolean): void {
           ~~~~~~


`;

exports[`static-const-name 1`] = `
error JSII8003: Static constant names must use ALL_CAPS, PascalCase, or camelCase. Rename "MyClass.snake_case" to "SNAKE_CASE"

`;

exports[`static-member-mixing.1 1`] = `
neg.static-member-mixing.1.ts:11:1 - error JSII5014: Instance member "funFunction" of class "jsii.Sub" conflicts with static member in ancestor "jsii.SuperDuper"

 11 export class Sub extends Super {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 12   public funFunction() {
    ~~~~~~~~~~~~~~~~~~~~~~~~
... 
 14   }
    ~~~
 15 }
    ~


`;

exports[`static-member-mixing.2 1`] = `
neg.static-member-mixing.2.ts:1:1 - error JSII5013: Member "funFunction" of class "jsii.TheClass" has both a static and an instance delcaration

  1 export class TheClass {
    ~~~~~~~~~~~~~~~~~~~~~~~
  2   public static funFunction() {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
... 
  8   }
    ~~~
  9 }
    ~


`;

exports[`static-method-name 1`] = `
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "jsii.MyClass.METHOD" to "method"

`;

exports[`static-method-name.1 1`] = `
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "jsii.MyClass.MethodIsNotCamelCase" to "methodIsNotCamelCase"

`;

exports[`static-prop-name.1 1`] = `
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "jsii.MyClass.Prop" to "prop"

`;

exports[`static-prop-name.2 1`] = `
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "jsii.MyClass.PROP" to "prop"

`;

exports[`struct-extends-interface 1`] = `
neg.struct-extends-interface.ts:6:18 - error JSII8007: Interface contains behavior. Rename "Struct" to "IStruct"

6 export interface Struct extends IInterface {
                   ~~~~~~


`;

exports[`submodules-cannot-have-colliding-names 1`] = `
neg.submodules-cannot-have-colliding-names.ts:3:14 - error JSII5011: Submodule "ns1" conflicts with "Ns1, as different languages could represent it as: "ns1", "Ns1""

3 export class Ns1 {
               ~~~

  neg.submodules-cannot-have-colliding-names.ts:1:13
    1 export * as ns1 from './namespaced';
                  ~~~
    This is the conflicting submodule declaration

`;

exports[`submodules-cannot-share-symbols 1`] = `
namespaced/index.ts:1:14 - error JSII3003: Symbol is exported under two distinct submodules: ns1 and ns2

1 export class Declaration {
               ~~~~~~~~~~~

  neg.submodules-cannot-share-symbols.ts:1:8
    1 export * as ns1 from './namespaced';
             ~~~~~~~~
    Symbol is exported under the "ns1" submodule
  neg.submodules-cannot-share-symbols.ts:2:8
    2 export * as ns2 from './namespaced';
             ~~~~~~~~
    Symbol is exported under the "ns2" submodule

`;

exports[`submodules-must-be-camel-cased 1`] = `
neg.submodules-must-be-camel-cased.ts:1:13 - error JSII8004: Submodule namespaces must be camelCased or snake_cased. Rename "Ns1" to ns1

1 export * as Ns1 from './namespaced';
              ~~~


`;
